xarray/tests/test_async.py

```diff
@@ -0,0 +1,196 @@
+                Store to wrap
+            latency : float
+                Amount of artificial latency to add to each get call, in seconds.
+            """
+            super().__init__(store)
+            self.latency = latency
+
+        def __str__(self) -> str:
+            return f"latency-{self._store}"
+
+        def __repr__(self) -> str:
+            return f"LatencyStore({self._store.__class__.__name__}, '{self._store}', latency={self.latency})"
+
+        async def get(
+            self,
+            key: str,
+            prototype: BufferPrototype,
+            byte_range: ByteRequest | None = None,
+        ) -> Buffer | None:
+            await asyncio.sleep(self.latency)
+            return await self._store.get(
+                key=key, prototype=prototype, byte_range=byte_range
+            )
+
+        async def get_partial_values(
+            self,
+            prototype: BufferPrototype,
+            key_ranges: Iterable[tuple[str, ByteRequest | None]],
+        ) -> list[Buffer | None]:
+            await asyncio.sleep(self.latency)
+            return await self._store.get_partial_values(
+                prototype=prototype, key_ranges=key_ranges
+            )
+else:
+    LatencyStore = {}
+
+
+@pytest.fixture
+def memorystore() -> "MemoryStore":
+    memorystore = zarr.storage.MemoryStore({})
+    z1 = zarr.create_array(
+        store=memorystore,
+        name="foo",
+        shape=(10, 10),
+        chunks=(5, 5),
+        dtype="f4",
+        dimension_names=["x", "y"],
+    )
+    z1[:, :] = np.random.random((10, 10))
+
+    z2 = zarr.create_array(
+        store=memorystore,
+        name="x",
+        shape=(10,),
+        chunks=(5),
+        dtype="f4",
+        dimension_names=["x"],
+    )
+    z2[:] = np.arange(10)
+
+    return memorystore
+
+
+class AsyncTimer:
+    """Context manager for timing async operations and making assertions about their execution time."""
+
+    start_time: float
+    end_time: float
+    total_time: float
+
+    @asynccontextmanager
+    async def measure(self):
+        """Measure the execution time of the async code within this context."""
+        self.start_time = time.time()
+        try:
+            yield self
+        finally:
+            self.end_time = time.time()
+            self.total_time = self.end_time - self.start_time
+
+
+@requires_zarr_v3
+@pytest.mark.asyncio
+class TestAsyncLoad:
+    LATENCY: float = 1.0
+
+    @pytest.fixture(params=["var", "ds", "da"])
+    def xr_obj(self, request, memorystore) -> xr.Dataset | xr.DataArray | xr.Variable:
+        latencystore = LatencyStore(memorystore, latency=self.LATENCY)
+        ds = xr.open_zarr(latencystore, zarr_format=3, consolidated=False, chunks=None)
+
+        match request.param:
+            case "var":
+                return ds["foo"].variable
+            case "da":
+                return ds["foo"]
+            case "ds":
+                return ds
+
+    def assert_time_as_expected(
```

Comment:
Let's instead use mocks to assert the async methods were called. Xarray's job is to do that only